File: 프로그래머스/lv2/42626. 더 맵게/더 맵게.java
import java.util.*;

class Solution {
    public int solution(int[] scoville, int K) {
        int answer = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>();
        
        for(int s : scoville) {
            pq.offer(s);
        }
        
        while(pq.peek() < K) {
            int first = pq.poll();
            if(pq.isEmpty()) return -1;
            int second = pq.poll();
            pq.offer(first + second*2);
            answer++;
        }
        
        // if(pq.isEmpty()) return -1;
        return answer;
    }
}

File: 프로그래머스/lv2/42626. 더 맵게/더 맵게.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    int newSco = -1;
    priority_queue<int, vector<int>, greater<int>> q;   // min heap
    
    for(int i=0; i<scoville.size(); i++)
        q.push(scoville[i]);
    
    while(q.top() < K) {
        newSco = q.top();
        q.pop();
        if(q.empty())return -1;
        newSco += (q.top() * 2);
        q.pop();
        q.push(newSco);
        answer++;
    }
    return answer;
    
}